### PR TITLE
Fix issue #1049 - Characteristic section is missing.

### PR DESCRIPTION
**Pull request by yjt98765**
## Main changes
- Add a CharacteristicSection member to CallableDeclaration;
- Add a toCharacteristicSection function and a CharacteristicVisitor type to build CharacteristicSection.

## Testing
By running `.\src\QsFmt\App\bin\Debug\netcoreapp3.1\qsfmt.exe src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestAliasCounts.qs` now, the output is
```
operation ApplyOp(dummy : Double, coefficients : Complex[], qubits : LittleEndian) : Unit is Adj + Ctl {
```
which is the same as the source file. Note that the `is Adj + Ctl` part is successfully printed.

In addition, I enabled the `Identity preserves original source code` test in `src\QsFmt\Formatter.Tests\FixedPoints.fs` by removing the `(Skip = "Not supported.")` part. (I added it back after the testing.) Before the modification, the test result is 
```
Failed!  - Failed:    52, Passed:    38, Skipped:     4, Total:    94, Duration: 1 s - Microsoft.Quantum.QsFmt.Formatter.Tests.dll (netcoreapp3.1)
```
After the modification, the test result is
```
Failed!  - Failed:    51, Passed:    39, Skipped:     4, Total:    94, Duration: 1 s - Microsoft.Quantum.QsFmt.Formatter.Tests.dll (netcoreapp3.1)
```
The failed and passed numbers change because a previously failed case now passes: `TestAliasCounts.qs`. I have verified it by running
```
Compare-Object -ReferenceObject (Get-Content -Path src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestAliasCounts.qs) -DifferenceObject (.\src\QsFmt\App\bin\Debug\netcoreapp3.1\qsfmt.exe src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestAliasCounts.qs)
```
There is no output, meaning the formatted program is the same as the original.

In conclusion, probably, I have fixed the problem reported in [issue 1049](https://github.com/microsoft/qsharp-compiler/issues/1049) without affecting other parts of the fomatter.

### CODE REVIEW

**Review by samarsha — COMMENTED** [review]
Thank you, it looks great! Just added a couple minor comments.

Could you add a test case or two to [Formatter.Tests/FixedPoints.fs](https://github.com/microsoft/qsharp-compiler/blob/main/src/QsFmt/Formatter.Tests/FixedPoints.fs)? As you noticed, the "Identity preserves original source code" test still has a lot of failures, but we can add some small tests specifically for the characteristics section.

**Comment by yjt98765** [comment]
> Thank you, it looks great! Just added a couple minor comments.
> 
> Could you add a test case or two to [Formatter.Tests/FixedPoints.fs](https://github.com/microsoft/qsharp-compiler/blob/main/src/QsFmt/Formatter.Tests/FixedPoints.fs)? As you noticed, the "Identity preserves original source code" test still has a lot of failures, but we can add some small tests specifically for the characteristics section.

Right. I have added two test cases in FixedPoints.fs regarding the characteristics section. I hope they meet the unit test standard.

File: src/QsFmt/Parser/QSharpParser.g4 [hunk 1]
```diff
@@ -74,14 +74,14 @@ parameter
...
-    : 'Adj'
-    | 'Ctl'
-    | '(' characteristicsExpression ')'
-    | characteristicsExpression '*' characteristicsExpression
-    | characteristicsExpression '+' characteristicsExpression
+    : 'Adj' # AdjointCharacteristics
+    | 'Ctl' # ControlledCharacteristics
+    | openParen='(' charExp=characteristicsExpression closeParen=')' # CharacteristicGroup
+    | left=characteristicsExpression '*' right=characteristicsExpression # AsteriskCharacteristics
+    | left=characteristicsExpression '+' right=characteristicsExpression # PlusCharacteristics
```
[the remark this answers is not in the document]
Comment (yjt98765) [reply src/QsFmt/Parser/QSharpParser.g4]: Glad to learn about this naming convention! I have updated these names in newly commits.

File: src/QsFmt/Formatter/ParseTree/Type.fs [hunk 2]
```diff
@@ -20,3 +55,15 @@ type TypeVisitor(tokens) =
...
+    let toCharacteristicSection tokens (context: QSharpParser.CharacteristicsContext) =
+        match context with
+        | null -> None
+        | context ->
+            Some(
+                {
+                    IsKeyword = context.is |> Node.toTerminal tokens
+                    Characteristic = (CharacteristicVisitor tokens).Visit context.charExp
+                }
+            )
```
[the remark this answers is not in the document]
Comment (yjt98765) [reply src/QsFmt/Formatter/ParseTree/Type.fs]: This would be much better! I did not know `Option.ofObj`. It seems very useful in this context. The code has been updated based on your suggestion.

**Review by samarsha — APPROVED** [review]
The tests and other changes look good, thanks for working on this!
